pkg/repo/etcd.go

```diff
@@ -4,32 +4,40 @@ import (
 	"context"
 	"fmt"
 	"log/slog"
+	"sync"
 	"time"
 
 	"go.etcd.io/etcd/clientv3"
 )
 
 type Etcd interface {
-	Put(ctx context.Context, cli *clientv3.Client, key string, value string) error
-	Client(cfg EtcdConfig) (*clientv3.Client, error)
+	Put(key string, value string) error
+
+	Watch(key string) (clientv3.WatchChan, context.CancelFunc)
+
+	Close() error
 }
 
 type EtcdHandler struct {
-	Conf *clientv3.Config
+	cli  *clientv3.Client
+	mu   *sync.RWMutex
+	ctx  context.Context
+	conf clientv3.Config
 }
 
-func (etcdHandler EtcdHandler) Client(cfg EtcdConfig) (*clientv3.Client, error) {
-	return clientv3.New(clientv3.Config{
-		Endpoints:   cfg.Servers,
-		DialTimeout: time.Duration(cfg.DialTimeoutSeconds * int(time.Second)),
-	})
+func (etcdHandler *EtcdHandler) Watch(key string) (clientv3.WatchChan, context.CancelFunc) {
+	watchCtx, cancel := context.WithCancel(etcdHandler.ctx)
+	watchRespChan := etcdHandler.cli.Watcher.Watch(watchCtx, key)
+
+	return watchRespChan, cancel
 }
 
-func (etcdHandler EtcdHandler) Put(ctx context.Context, cli *clientv3.Client, key, value string) error {
-	defer cli.Close()
+func (etcdHandler *EtcdHandler) Put(key, value string) error {
+	etcdHandler.mu.Lock()
+	defer etcdHandler.mu.Unlock()
 
-	ctx, cancel := context.WithTimeout(ctx, etcdHandler.Conf.DialTimeout)
-	res, err := cli.Put(ctx, key, value)
+	ctx, cancel := context.WithTimeout(etcdHandler.ctx, etcdHandler.conf.DialTimeout)
+	res, err := etcdHandler.cli.Put(ctx, key, value)
```

Comment:
I think etcd client should be thread safe. are you sure we need lock here?

Reply:
yes based on implementation it looks thread safe https://github.com/etcd-io/etcd/blob/client/v3.5.16/client/v3/client.go#L45

Reply:
@ksysoev but it only acquires lock for updating endpoints. The `Op` methods do not seem to really use the mutex

Reply:
@shan-96 grpc client is also thread safe :)

Reply:
idk when I was the implementation of PUT, GET in the KV interface. I really didnt find any usage of mutex. But yeah the documentation mentions clients are both thread safe and reusable. I will remove it from etcd